track_analyzing/track_matcher.cpp

```diff
@@ -0,0 +1,254 @@
+                                                          IndexGraph const & graph,
+                                                          IVehicleModel const & vehicleModel,
+                                                          NumMwmId mwmId)
+{
+  index.ForEachInRect(
+      [&](FeatureType const & ft) {
+        if (!ft.GetID().IsValid())
+          return;
+
+        if (ft.GetID().m_mwmId.GetInfo()->GetType() != MwmInfo::COUNTRY)
+          return;
+
+        if (!vehicleModel.IsRoad(ft))
+          return;
+
+        ft.ParseGeometry(FeatureType::BEST_GEOMETRY);
+
+        for (size_t segIdx = 0; segIdx + 1 < ft.GetPointsCount(); ++segIdx)
+        {
+          double const distance =
+              DistanceToSegment(ft.GetPoint(segIdx), ft.GetPoint(segIdx + 1), m_point);
+          if (distance < kMatchingRange)
+          {
+            AddCandidate(Segment(mwmId, ft.GetID().m_index, static_cast<uint32_t>(segIdx),
+                                 true /* forward */),
+                         distance, graph);
+
+            if (!vehicleModel.IsOneWay(ft))
+            {
+              AddCandidate(Segment(mwmId, ft.GetID().m_index, static_cast<uint32_t>(segIdx),
+                                   false /* forward */),
+                           distance, graph);
+            }
+          }
+        }
+      },
+      MercatorBounds::RectByCenterXYAndSizeInMeters(m_point, kMatchingRange),
+      scales::GetUpperScale());
+}
+
+void TrackMatcher::Step::FillCandidates(Step const & previousStep, IndexGraph & graph)
+{
+  vector<SegmentEdge> edges;
+
+  for (Candidate const & candidate : previousStep.m_candidates)
+  {
+    Segment const & segment = candidate.GetSegment();
+    m_candidates.emplace_back(segment, DistanceToSegment(segment, m_point, graph));
+
+    edges.clear();
+    graph.GetEdgeList(segment, true /* isOutgoing */, edges);
+
+    for (SegmentEdge const & edge : edges)
+    {
+      Segment const & target = edge.GetTarget();
+      if (!segment.IsInverse(target))
+        m_candidates.emplace_back(target, DistanceToSegment(target, m_point, graph));
+    }
+  }
+
+  my::SortUnique(m_candidates);
+
+  m_candidates.erase(remove_if(m_candidates.begin(), m_candidates.end(),
+                               [&](Candidate const & candidate) {
+                                 return candidate.GetDistance() > kMatchingRange;
+                               }),
+                     m_candidates.end());
+}
+
+void TrackMatcher::Step::ChooseSegment(Step const & nextStep, IndexGraph & indexGraph)
+{
+  CHECK(!m_candidates.empty(), ());
+
+  double minDistance = numeric_limits<double>::max();
+
+  vector<SegmentEdge> edges;
+  indexGraph.GetEdgeList(nextStep.m_segment, false, edges);
+  edges.emplace_back(nextStep.m_segment, 0.0 /* weight */);
+
+  for (Candidate const & candidate : m_candidates)
+  {
+    if (candidate.GetDistance() < minDistance && EdgesContain(edges, candidate.GetSegment()))
```

Comment:
Why do you check that these is at least one edge from the next step that is ingoing to the candidate segment? Why not to check that there's at least one edge from the next step that is outgoing from the candidate segment?